### DISCUSSION TITLE
All parameters in a separate permanent file?

### DISCUSSION THREAD

**Discussion by corneel27** [Louisvdw/dbus-serialbattery#232] [Ideas]
I have installed V2.91 van VenusOS and V0.13 of this driver.
All runs well (2 JDB BMS 16 s, 100 A, with two USB RS485) and I am very happy with with this driver.
This evening I noticed that the current is limited when the SOC is <= 30%
I have seen the following part of code in battery.py:
```
        # Discharge depending on the SOC values
        if self.soc <= 10:
            self.control_discharge_current = 5
        elif 10 < self.soc <= 20:
            self.control_discharge_current = self.max_battery_discharge_current/4
        elif 20 < self.soc <= 30:
            self.control_discharge_current = self.max_battery_discharge_current/2
        else:
            self.control_discharge_current = self.max_battery_discharge_current
```
For me those 30% is to high.
I would rather think to start with limiting at 15%
I can change this for myself and each time when a new version is released I have a list of changes to work through.
My first proposal is for the next version:
Set al the used constants from utils.py and battery.py in a json file (for instance "config.json")
Deliver with the code a file (config.org) with all the current settings
as part of the installation script: 
        if config.json doesn't exist copy config.org to config.json
        else don't copy and the existing config.json remains.

A second proposal (if possible):
check every minute whether config.json is changed (timestamp of the file) and if so: restart the driver

I would like to contribute in the development of the driver with these enhancements.

Should I make a fork, apply these enhancements and then make a pullrequest?

**Comment by Louisvdw** [comment]
Hi
I'm glad the driver is working well for you.
I'm busy moving all the settings to be adjustable in the GUI. These settigs for the CCCM will be part of that also. 

If you want to help and make variables in the utils.py that for the limits that would be great. I'll then move those to the GUI when we get to them. Yes, just make a fork, make a new branch and then make a pull request from that branch with the changes.

**Comment by corneel27** [comment]
Ok so for a good understanding and to make no mistakes:
I will make constants in utils.py for the soc percentages and the factor to limit to the charge current.
Also for the discharge current?

**Comment by Louisvdw** [comment]
We should add this to both charge and discharge. I have created a ticket which should clarify this a bit showing each variable that we will need. If you add them in utils.py then they should slot in nicely with the code that handles currently. 
https://github.com/Louisvdw/dbus-serialbattery/issues/233

If you want to look at something else as well there is a ticket to split the enable/disable of the CCCM for charge and discharge limits so that it is not a both or nothing. For that the current flag in utils.py CCCM_ENABLE will only disable the charge part, and then we need a new variable DCCM_ENABLE to do the Discharge limit flag.
https://github.com/Louisvdw/dbus-serialbattery/issues/211

**Comment by corneel27** [comment]
Hello Louis,
Last few days I get warnings "battery to low"when the soc is below 20% and alarmswhen soc is below 10%
I investigated the code and I saw in the code for the JBD battery following lines (93, 94):
```
        # Software implementations for low soc
        self.protection.soc_low = 2 if self.soc < 10 else 1 if self.soc < 20 else 0
```
Do you have good reasons not to good below 20%?
Otherwise I would like to go till 10% and then those boundaries could be set in utils.py.
What is your opinion?
Cheers Cees

**Reply by Louisvdw** [reply]
Hi Cees
Yes those boundaries could also be set in utils.py for easy edits by users.
Most manufacturers (of batteries) only allow warrantees if the battery does not go too low and those low values are in most cases between 10%-20%. That is why the driver starts warning at 20% and then alarm below 10%.

**Comment by corneel27** [comment]
I made the requested constants for charge and discharge limits in utils.py.
I implemented these constants in battery.py
There is a pull request waiting for your review.

I have a question about the  "low soc"  warning in lltjbd.py:
Isn't it better to move this test to battery.py, because it is test which must be made for every type of battery?

**Comment by corneel27** [comment]
Hi Louis,
Can I help you with the implementation of the virtual (summed, total) battery?
Or perhaps other things?
Greetings Cees
